pkg/repository/repoprovider/unified_repo.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright 2020 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repoprovider
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/repository/repoconfig"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/util/ownership"
+)
+
+type unifiedRepoProvider struct {
+	ctx                  context.Context
+	credentialsFileStore credentials.FileStore
+	backupLocation       *velerov1api.BackupStorageLocation
+	repoService          udmrepo.BackupRepoService
+	repoPassword         string
+	repoName             string
+	configPath           string
+	log                  *logrus.Logger
+}
+
+// this func is assigned to a package-level variable so it can be
+// replaced when unit-testing
+var getAzureCredentials = repoconfig.GetAzureCredentials
+var getS3Credentials = repoconfig.GetS3Credentials
+var getGCPCredentials = repoconfig.GetGCPCredentials
+var getS3BucketRegion = repoconfig.GetAWSBucketRegion
+var getAzureStorageDomain = repoconfig.GetAzureStorageDomain
+
+func NewUnifiedRepoProvider(
+	ctx context.Context,
+	credentialFileStore credentials.FileStore,
+	backupLocation *velerov1api.BackupStorageLocation,
+	configPath string,
+	repoName string,
+	log *logrus.Logger,
+) (RepositoryProvider, error) {
+	repo := unifiedRepoProvider{
+		ctx:                  ctx,
+		credentialsFileStore: credentialFileStore,
+		backupLocation:       backupLocation,
+		configPath:           configPath,
+		repoName:             repoName,
+		log:                  log,
+	}
+
+	repo.repoService = createRepoService(log)
+
+	log.Debug("Finished create unified repo service")
+
+	return &repo, nil
+}
+
+func (urp *unifiedRepoProvider) InitRepo(ctx context.Context, backupLocation string) error {
+	log := urp.log.WithFields(logrus.Fields{
+		"backupLocation": backupLocation,
+	})
+
+	log.Debug("Start to init repo")
+
+	err := urp.ensureRepoPassword()
+	if err != nil {
+		log.WithError(err).Error("Failed to ensure repo password")
```

Comment:
Don't log the error here, just wrap it and return it directly, the caller should log the error if necessary, or there will be duplicated error message in the log

Reply:
OK, will remove all the logs for the errors.

Reply:
Done